MC/analysis_testing/o2dpg_analysis_test_workflow.py

```diff
@@ -149,50 +275,58 @@ def add_analysis_qc_upload_tasks(workflow, prodcution_tag, run_number, *ana_task
         ana_tasks_expected_outputs: list of tuples
             [(AnalysisName_1, (expected_output_1_1, expected_output_1_2, ...)), ..., (AnalysisName_N, (expected_output_N_1, expected_output_N_2, ...)) ]
     """
-    if not ana_tasks_expected_outputs:
-        ana_tasks_expected_outputs = DEFAULT_ANALYSES_FOR_QC_UPLOAD
+    analyses_to_add_for = {}
+    # collect analyses in current workflow
+    for task in workflow:
+        if ANALYSIS_LABEL in task["labels"] or ANALYSIS_LABEL_MERGED in task["labels"]:
+            analyses_to_add_for[task["name"]] = task
 
-    for ana_name_raw, expcted_outputs in ana_tasks_expected_outputs:
+    for ana in ANALYSES:
+        if not ana["expected_output"]:
+            continue
+        ana_name_raw = ana["name"]
         ana_name = full_ana_name(ana_name_raw)
-        for pot_ana in workflow:
-            # search through workflow stages if we can find the requested analysis
-            if pot_ana["name"] != ana_name:
-                continue
-            print(f"Adding QC upload task for analysis {ana_name_raw}")
-            cwd = pot_ana["cwd"]
-            qc_tag = f"Analysis{ana_name_raw}"
-            needs = [ana_name]
-            for eo in expcted_outputs:
-                # this seems unnecessary but to ensure backwards compatible behaviour...
-                rename_output = eo.strip(".root")
-                rename_output = f"{rename_output}_{ana_name_raw}.root"
-                # add upload task for each expected output file
-                task = createTask(name=f"{ANALYSIS_LABEL}_finalize_{ana_name_raw}_{rename_output}", cwd=cwd, lab=[f"{ANALYSIS_LABEL}Upload", ana_name_raw], cpu=1, mem='2000', needs=needs)
-                # This has now to be renamed for upload, as soon as that is done, the output is renamed back to its original, there is in general no point of renaming it on disk only because one specific tasks needs a renamed version of it
-                rename_cmd = f"mv {eo} {rename_output}"
-                rename_back_cmd = f"mv {rename_output} {eo}"
-                task["cmd"] = f"{rename_cmd} && o2-qc-upload-root-objects --input-file ./{rename_output} --qcdb-url ccdb-test.cern.ch:8080 --task-name Analysis{ana_name_raw} --detector-code AOD --provenance qc_mc --pass-name passMC --period-name {prodcution_tag} --run-number {run_number} && {rename_back_cmd} "
-                workflow.append(task)
+        if ana_name not in analyses_to_add_for:
+            continue
+        # search through workflow stages if we can find the requested analysis
+        print(f"Adding QC upload task for analysis {ana_name_raw}")
+        pot_ana = analyses_to_add_for[ana_name]
+        cwd = pot_ana["cwd"]
+        qc_tag = f"Analysis{ana_name_raw}"
+        needs = [ana_name]
+        provenance, pass_name = ("qc_mc", "passMC") if ANALYSIS_LABEL_ON_MC in pot_ana["labels"] else ("qc_data", "pass_data")
```

Comment:
Please use `"qc_async"` instead of `"qc_data"`. This is what we use for QC results based on real data reconstructed asynchronously.

Reply:
Then, in case of data reconstruction, `pass_name` usually comes from grid JDLs, so perhaps one could expose it as an argument. Usually, the convention `<period_name>_<pass_name>` is used in production tags, so alternatively we could extract it from there, if we trust it enough.

Reply:
@knopers8 I applied the changes as requested, you can have a look.

Reply:
thanks, it seems all good now!